janusgraph-core/src/main/java/org/janusgraph/diskstorage/keycolumnvalue/scan/MultiThreadsScanner.java

```diff
@@ -0,0 +1,259 @@
+        this.graphConfiguration = graphConfiguration;
+
+        this.dataQueues = new ArrayList<>(queries.size());
+        this.pullThreads = new ArrayList<>(queries.size());
+
+        setUp(queries);
+    }
+
+    private void setUp(List<SliceQuery> queries) throws BackendException {
+        int numQueries = queries.size();
+
+        for (int pos = 0; pos < numQueries; pos++) {
+            pullThreads.add(addDataPuller(queries.get(pos), storeTx, pos));
+        }
+    }
+
+    void run() throws InterruptedException, TemporaryBackendException {
+        int numQueries = queries.size();
+        SliceResult[] currentResults = new SliceResult[numQueries];
+        while (!interrupted) {
+            collectDataFromPullers(currentResults, numQueries);
+
+            SliceResult conditionQuery = currentResults[0];
+            if (conditionQuery==null) break; //Termination condition - primary query has no more data
+            final StaticBuffer key = conditionQuery.key;
+
+            Row e = buildRow(numQueries, currentResults, key);
+
+            rowQueue.put(e);
+        }
+    }
+
+    private void collectDataFromPullers(SliceResult[] currentResults, int numQueries) throws InterruptedException, TemporaryBackendException {
+        for (int i = 0; i < numQueries; i++) {
+            if (currentResults[i]!=null) continue;
+            BlockingQueue<SliceResult> queue = dataQueues.get(i);
+
+            SliceResult qr = queue.poll(TIME_PER_TRY, TimeUnit.MILLISECONDS); //Try very short time to see if we are done
+            if (qr==null) {
+                DataPuller dataPuller = pullThreads.get(i);
+                if (dataPuller.isFinished()) continue; //No more data to be expected
+                while (!dataPuller.isFinished() && qr == null) {
+                    qr = queue.poll(TIME_PER_TRY, TimeUnit.MILLISECONDS);
+                }
+                if (qr==null && !dataPuller.isFinished())
+                    throw new TemporaryBackendException("Timed out waiting for next row data - storage error likely");
+            }
+            currentResults[i]=qr;
+        }
+    }
+
+    private Row buildRow(int numQueries, SliceResult[] currentResults, StaticBuffer key) {
+        Map<SliceQuery,EntryList> queryResults = new HashMap<>(numQueries);
+        for (int i = 0; i< currentResults.length; i++) {
+            SliceQuery query = queries.get(i);
+            EntryList entries = EntryList.EMPTY_LIST;
+            if (currentResults[i]!=null && currentResults[i].key.equals(key)) {
+                assert query.equals(currentResults[i].query);
+                entries = currentResults[i].entries;
+                currentResults[i]=null;
+            }
+            queryResults.put(query,entries);
+        }
+        return new Row(key, queryResults);
+    }
+
+    @Override
+    void join() throws InterruptedException {
+        int i = 0;
+        for (DataPuller dataPuller : pullThreads) {
+            dataPuller.join(10);
```

Comment:
(nitpick) I would recommend placing such constants on top of the class to have better observability of hard-coded constants. That said, I see that this part was already in the previous code, so not necessary to resolve.